sycl/doc/SYCLNativeCPU.md

```diff
@@ -0,0 +1,170 @@
+* Allow the SYCL runtime to call the kernel, by registering it to the SYCL runtime, operation performed by the [EmitSYCLNativeCPUHeader pass](llvm/lib/SYCLLowerIR/EmitSYCLNativeCPUHeader.cpp).
+
+
+## PrepareSYCLNativeCPU Pass
+
+This pass will add a pointer to a `nativecpu_state` struct as kernel argument to all the kernel functions, and it will replace all the uses of SPIRV builtins with the return value of appropriately defined functions, which will read the requested information from the `nativecpu_state` struct. The `nativecpu_state` struct and the builtin functions are defined in [native_cpu.hpp](sycl/include/sycl/detail/native_cpu.hpp).
+
+
+The resulting IR is:
+
+```llvm
+define weak dso_local void @_Z6Sample(ptr noundef align 4 %0, ptr noundef align 4 %1, ptr noundef align 4 %2, ptr %3) local_unnamed_addr #3 !srcloc !74 !kernel_arg_buffer_location !75 !kernel_arg_type !76 !sycl_fixed_targets !49 !sycl_kernel_omit_args !77 {
+entry:
+  %ncpu_builtin = call ptr @_Z13get_global_idmP15nativecpu_state(ptr %3)
+  %4 = load i64, ptr %ncpu_builtin, align 32, !noalias !78
+  %arrayidx.i = getelementptr inbounds i32, ptr %1, i64 %4
+  %5 = load i32, ptr %arrayidx.i, align 4, !tbaa !72
+  %arrayidx4.i = getelementptr inbounds i32, ptr %2, i64 %4
+  %6 = load i32, ptr %arrayidx4.i, align 4, !tbaa !72
+  %add.i = add nsw i32 %5, %6
+  %cmp.i8.i = icmp ult i64 %4, 2147483648
+  tail call void @llvm.assume(i1 %cmp.i8.i)
+  %arrayidx6.i = getelementptr inbounds i32, ptr %0, i64 %4
+  store i32 %add.i, ptr %arrayidx6.i, align 4, !tbaa !72
+  ret void
+}
+```
+This pass will also set the correct calling convention for the target, and handle calling convention-related function attributes, allowing to call the kernel from the runtime.
+
+## EmitSYCLNativeCPUHeader pass
+
+This pass emits an additional integration header, that will be compiled by the host compiler during the host compilation step. This header is included by the main integration footer and does not need to be managed manually. Its main purpose is to enable the SYCL runtime to register kernels and to call kernels that had unused parameters removed by the optimizer. The header contains, for each kernel:
+* The kernel declaration as a C++ function, all pointer arguments are emitted as `void *`, the scalar arguments maintain their type.
+* A `subhandler` definition, which unpacks the vector of kernel arguments coming from the SYCL runtime, and forwards only the used arguments to the kernel.
+* The definition of `_pi_offload_entry_struct`, `pi_device_binary_struct` and `pi_device_binaries_struct` variables, and a call to `__sycl_register_lib`, which allows to register the kernel to the sycl runtime (the call to `__sycl_register_lib` is performed at program startup via the constructor of a global). The Native CPU integration header is always named `<main-sycl-int-header>.hc`.
```

Comment:
Thanks for the comments. We are planning to integrate with the offload_wrapper_tool [(See todo comment)](https://github.com/PietroGhg/llvm/blob/650092cd4c3118b6b9f5ecf896247c333866508e/llvm/lib/SYCLLowerIR/EmitSYCLNativeCPUHeader.cpp#L166)  which would need to know if unused kernel arguments were removed by the optimizer. Once we have a solution for this we wanted to create another PR and replace the current handler function implementation from this PR. We'll list this as ongoing work.

The `native_cpu` target should be able to coexist with other targets - like other AOT targets at the expense of compile time and file size. The plan is to support this in a subsequent PR.

Reply:
If the direction to align designs and make `native_cpu` co-exist with other targets is explicitly noted in docs, I'm fine with the current approach.

However, from a [linked comment](https://github.com/PietroGhg/llvm/blob/650092cd4c3118b6b9f5ecf896247c333866508e/llvm/lib/SYCLLowerIR/EmitSYCLNativeCPUHeader.cpp#L166) it seems like the main blocker is dead kernel argument elimination optimization, which is not required by core SYCL spec. Would it be better to say that if `native_cpu` is one of a targets, then the option can't be used and optimization had to be switched off? This would allow us to converge implementations earlier, without introducing soon-to-be-legacy code at all.

Are there other roadblocks?

Reply:
We've updated the design docs to mention coexistence of `native_cpu` with other targets as a TODO. We believe this coexistence is possible even with the current approach, we just haven't prioritized it yet.

Aligning with `clang-offload-wrapper` was also lower priority because the current approach was quicker for us to get working. Temporarily disabling the removal of unused kernel arguments would address the issue in the [linked comment](https://github.com/PietroGhg/llvm/blob/650092cd4c3118b6b9f5ecf896247c333866508e/llvm/lib/SYCLLowerIR/EmitSYCLNativeCPUHeader.cpp#L166), but the alignment will likely require additional changes. We don't expect any showstoppers, it's just a matter of prioritizing and right now we'd like to complete SYCL feature support first. For the alignment with `clang-offload-wrapper` we expect to reuse a lot of work from the current approach so hopefully not much will turn into legacy code.

Reply:
Marking this as resolved for now.

About `clang-offload-wrapper`: it seems to me that the whole `EmitSYCLNativeCPUHeader` pass can be replaced with a custom path in `clang-offload-wrapper` (or upcoming `clang-linker-wrapper`): instead of embedding the whole input module into a global variable as a blob, the tool would instead modify input module to provide necessary "interface" functions, renaming kernels to avoid conflicts, emitting global constructors to register them, etc.